### PR TITLE
Ensure all data state stores save is encrypted 

### LINKED ISSUE #609: sled-store: Encrypt all data in store
**Is your feature request related to a problem? Please describe.**
Currently some information isn't encrypted in the sled store (probably also other store implementation). This way an adversary who gains access to the files on disk can extract some information about the users communication. 

**Describe the solution you'd like**
Encrypt all data and metadata. Probably it would be good to implement an additional layer that encrypts the local files instead of encrypting the data on a value base.  So that it's not possible to forget to encrypt an entry in the db.

**Additional context**
As context, an security audit is performed for Fractal and some of it's dependencies. The results will be published once the audit is completed.

### PR DESCRIPTION
**Pull request by gnunicorn**
... if a cipher is provided.

Effected:

- [x] sled state store
   - [x] filters
   - [x] session
   - [x] media
   - [x] custom
   - [x] [invited|joined]_user_ids
   - [x] room_timeline_metadata 
- [x] indexeddb state store
   - [x] filters
   - [x] room_timeline_metadata 

every other instance of `insert` (in `sled`) and `set_key_val` (in `indexeddb`) has a proper key and value serialization as far as I could see.
 

fixes #609 and brings some more integration tests to cover these functions (though whether they en- & decrypt is black boxed and thus not something we can tests for :( )

### CODE REVIEW

File: crates/matrix-sdk-base/src/store/integration_tests.rs [hunk 1]
```diff
@@ -358,6 +358,16 @@ macro_rules! statestore_integration_tests {
                     assert!(!members.is_empty(), "We expected to find members for the room")
                 }
 
+                #[async_test]
+                async fn test_filter_saving() {
+                    let store = get_store().await.unwrap();
+                    let test_name = "filter_name";
+                    let filter_id = "filter_id_1234";
+                    assert_eq!(store.get_filter(test_name).await.unwrap(), None);
+                    assert_eq!(store.save_filter(test_name, filter_id).await.unwrap(), ());
```
Comment (jplatte) [inline crates/matrix-sdk-base/src/store/integration_tests.rs:367]: Assertion is a no-op.

File: crates/matrix-sdk-base/src/store/integration_tests.rs [hunk 2]
```diff
@@ -368,6 +368,19 @@ macro_rules! statestore_integration_tests {
                     assert_eq!(store.get_filter(test_name).await.unwrap(), Some(filter_id.to_owned()));
                 }
 
+                #[async_test]
+                async fn test_sync_token_saving() {
+
```
Comment (jplatte) [inline crates/matrix-sdk-base/src/store/integration_tests.rs:373]: Extra newline should be removed.

File: crates/matrix-sdk-base/src/store/integration_tests.rs [hunk 3]
```diff
@@ -368,6 +368,19 @@ macro_rules! statestore_integration_tests {
...
+                    let mut changes = StateChanges::default();
+                    let store = get_store().await.unwrap();
+                    let sync_token = "t392-516_47314_0_7_1".to_owned();
+
+                    changes.sync_token = Some(sync_token.clone());
+                    assert_eq!(store.get_sync_token().await.unwrap(), None);
+                    assert_eq!(store.save_changes(&changes).await.unwrap(), ());
```
Comment (jplatte) [inline crates/matrix-sdk-base/src/store/integration_tests.rs:380]: Another no-op assertion.

File: crates/matrix-sdk-sled/src/encode_key.rs [hunk 4]
```diff
@@ -39,6 +39,12 @@ impl<T: EncodeKey + ?Sized> EncodeKey for &T {
     }
 }
 
+impl EncodeKey for &[u8] {
+    fn encode_as_bytes(&self) -> Cow<'_, [u8]> {
+        (*self).into()
```
Comment (jplatte) [inline crates/matrix-sdk-sled/src/encode_key.rs:44]: Do we really want this impl? The slice could contain our `ENCODE_SEPARATOR`.

**Review by jplatte — COMMENTED** [review]
Small nitpicks about commit messages: "also" seems unnecessary, and "cleaning up" should be "clean up". Additionally, the commit messages are inconsistent about capitalization, would be nice to always start uppercase after the first colon.

File: crates/matrix-sdk-sled/src/encode_key.rs (hunk 4 above)
Comment (gnunicorn) [reply crates/matrix-sdk-sled/src/encode_key.rs:44]: hmm.... while that is a fair argument in general, this is only used (and usable for the outside) via the `custom`-feature-are and there we are a) in our own bucket and b) don't (yet) use the encode separator. 

However, I agree that this is a very generic implementation and not safe as it is. I will give it a wrapper type that one has to use instead.

**Comment by jplatte** [comment]
Test failure seems legit (not spurious):

```
---- state_store::encrypted_tests::integration::test_room_timeline stdout ----
thread 'state_store::encrypted_tests::integration::test_room_timeline' panicked at 'called `Result::unwrap()` on an `Err` value: Json(Error("missing field `start`", line: 1, column: 724))', crates/matrix-sdk-sled/src/state_store.rs:1609:5
```

File: crates/matrix-sdk-sled/src/state_store.rs [hunk 5]
```diff
@@ -398,22 +399,22 @@ impl SledStore {
     }
 
     pub async fn save_filter(&self, filter_name: &str, filter_id: &str) -> Result<()> {
-        self.session.insert(self.encode_key(SESSION, ("filter", filter_name)), filter_id)?;
+        self.session.insert(
+            self.encode_key(SESSION, ("filter", filter_name)),
+            self.serialize_value(&filter_id)?,
```
Comment (gnunicorn) [inline crates/matrix-sdk-sled/src/state_store.rs:404]: this changes the layout of the store, we should bump its version number

**Pull request merged**